From fb569856df6ccc6ad14e191294c568968ed4a99f Mon Sep 17 00:00:00 2001
From: Erhan Bilgili <erhan.bilgili@gmail.com>
Date: Fri, 10 Jul 2026 19:43:36 +0200
Subject: [PATCH] win32u: Avoid full hit-test list for mouse tracking.

(cherry picked from commit fb569856df6ccc6ad14e191294c568968ed4a99f)

Source: https://github.com/nanomatters/wine-salkim/commit/fb569856df6ccc6ad14e191294c568968ed4a99f
---
 dlls/win32u/window.c           | 31 +++++++++++++++++++++++++++++
 include/wine/server_protocol.h | 22 ++++++++++++++++++++-
 server/protocol.def            | 11 +++++++++++
 server/request_handlers.h      | 10 ++++++++++
 server/request_trace.h         | 17 ++++++++++++++++
 server/window.c                | 36 ++++++++++++++++++++++++++++++++++
 6 files changed, 126 insertions(+), 1 deletion(-)

diff --git a/dlls/win32u/window.c b/dlls/win32u/window.c
index 406b30c36e0..cf346dc9020 100644
--- a/dlls/win32u/window.c
+++ b/dlls/win32u/window.c
@@ -2917,6 +2917,27 @@ static HWND *list_children_from_point( HWND hwnd, POINT pt, UINT dpi )
     return NULL;
 }
 
+static HWND first_window_from_point( HWND hwnd, POINT pt, UINT dpi, DWORD *style )
+{
+    HWND ret = 0;
+
+    SERVER_START_REQ( get_window_from_point )
+    {
+        req->parent = wine_server_user_handle( hwnd );
+        req->x = pt.x;
+        req->y = pt.y;
+        req->dpi = dpi;
+        if (!wine_server_call( req ))
+        {
+            ret = wine_server_ptr_handle( reply->handle );
+            *style = reply->style;
+        }
+    }
+    SERVER_END_REQ;
+
+    return ret;
+}
+
 static HWND window_from_point_excluding_root( HWND hwnd, POINT pt )
 {
     HWND hwnd_root = NtUserGetAncestor( hwnd, GA_ROOT );
@@ -2958,6 +2979,16 @@ HWND window_from_point( HWND hwnd, POINT pt, INT *hittest, BOOL send_nchittest )
 
     *hittest = HTNOWHERE;
 
+    if (!send_nchittest)
+    {
+        DWORD style;
+
+        if (!(ret = first_window_from_point( hwnd, pt, dpi, &style ))) return 0;
+        *hittest = (style & WS_DISABLED) ? HTERROR : HTCLIENT;
+        TRACE( "scope %p (%d,%d) returning %p\n", hwnd, pt.x, pt.y, ret );
+        return ret;
+    }
+
     if (!(list = list_children_from_point( hwnd, pt, dpi ))) return 0;
 
     /* now determine the hittest */
diff --git a/include/wine/server_protocol.h b/include/wine/server_protocol.h
index cc9d09937ff..235f5e35e37 100644
--- a/include/wine/server_protocol.h
+++ b/include/wine/server_protocol.h
@@ -3815,6 +3815,23 @@ struct get_window_children_from_point_reply
 };
 
 
+struct get_window_from_point_request
+{
+    struct request_header __header;
+    user_handle_t  parent;
+    int            x;
+    int            y;
+    int            dpi;
+    char __pad_28[4];
+};
+struct get_window_from_point_reply
+{
+    struct reply_header __header;
+    user_handle_t  handle;
+    unsigned int   style;
+};
+
+
 
 struct get_window_tree_request
 {
@@ -6630,6 +6647,7 @@ enum request
     REQ_get_window_list,
     REQ_get_class_windows,
     REQ_get_window_children_from_point,
+    REQ_get_window_from_point,
     REQ_get_window_tree,
     REQ_set_window_pos,
     REQ_get_window_rectangles,
@@ -6955,6 +6973,7 @@ union generic_request
     struct get_window_list_request get_window_list_request;
     struct get_class_windows_request get_class_windows_request;
     struct get_window_children_from_point_request get_window_children_from_point_request;
+    struct get_window_from_point_request get_window_from_point_request;
     struct get_window_tree_request get_window_tree_request;
     struct set_window_pos_request set_window_pos_request;
     struct get_window_rectangles_request get_window_rectangles_request;
@@ -7278,6 +7297,7 @@ union generic_reply
     struct get_window_list_reply get_window_list_reply;
     struct get_class_windows_reply get_class_windows_reply;
     struct get_window_children_from_point_reply get_window_children_from_point_reply;
+    struct get_window_from_point_reply get_window_from_point_reply;
     struct get_window_tree_reply get_window_tree_reply;
     struct set_window_pos_reply set_window_pos_reply;
     struct get_window_rectangles_reply get_window_rectangles_reply;
@@ -7442,6 +7462,6 @@ union generic_reply
     struct hwnd_dmabuf_release_channel_reply hwnd_dmabuf_release_channel_reply;
 };
 
-#define SERVER_PROTOCOL_VERSION 935
+#define SERVER_PROTOCOL_VERSION 936
 
 #endif /* __WINE_WINE_SERVER_PROTOCOL_H */
diff --git a/server/protocol.def b/server/protocol.def
index 312aa1dc28d..3a4fc1ea4af 100644
--- a/server/protocol.def
+++ b/server/protocol.def
@@ -2868,6 +2868,17 @@ enum message_type
     VARARG(children,user_handles); /* children handles */
 @END
 
+/* Get the first window that contains a given point */
+@REQ(get_window_from_point)
+    user_handle_t  parent;        /* parent window */
+    int            x;             /* point in parent coordinates */
+    int            y;
+    int            dpi;           /* dpi for the point coordinates */
+@REPLY
+    user_handle_t  handle;        /* window handle */
+    unsigned int   style;         /* window style */
+@END
+
 
 /* Get window tree information from a window handle */
 @REQ(get_window_tree)
diff --git a/server/request_handlers.h b/server/request_handlers.h
index c3c6e93e17a..8c0a1b884a9 100644
--- a/server/request_handlers.h
+++ b/server/request_handlers.h
@@ -163,6 +163,7 @@ DECL_HANDLER(get_window_parents);
 DECL_HANDLER(get_window_list);
 DECL_HANDLER(get_class_windows);
 DECL_HANDLER(get_window_children_from_point);
+DECL_HANDLER(get_window_from_point);
 DECL_HANDLER(get_window_tree);
 DECL_HANDLER(set_window_pos);
 DECL_HANDLER(get_window_rectangles);
@@ -485,6 +486,7 @@ static const req_handler req_handlers[REQ_NB_REQUESTS] =
     (req_handler)req_get_window_list,
     (req_handler)req_get_class_windows,
     (req_handler)req_get_window_children_from_point,
+    (req_handler)req_get_window_from_point,
     (req_handler)req_get_window_tree,
     (req_handler)req_set_window_pos,
     (req_handler)req_get_window_rectangles,
@@ -1594,6 +1596,14 @@ C_ASSERT( offsetof(struct get_window_children_from_point_request, dpi) == 24 );
 C_ASSERT( sizeof(struct get_window_children_from_point_request) == 32 );
 C_ASSERT( offsetof(struct get_window_children_from_point_reply, count) == 8 );
 C_ASSERT( sizeof(struct get_window_children_from_point_reply) == 16 );
+C_ASSERT( offsetof(struct get_window_from_point_request, parent) == 12 );
+C_ASSERT( offsetof(struct get_window_from_point_request, x) == 16 );
+C_ASSERT( offsetof(struct get_window_from_point_request, y) == 20 );
+C_ASSERT( offsetof(struct get_window_from_point_request, dpi) == 24 );
+C_ASSERT( sizeof(struct get_window_from_point_request) == 32 );
+C_ASSERT( offsetof(struct get_window_from_point_reply, handle) == 8 );
+C_ASSERT( offsetof(struct get_window_from_point_reply, style) == 12 );
+C_ASSERT( sizeof(struct get_window_from_point_reply) == 16 );
 C_ASSERT( offsetof(struct get_window_tree_request, handle) == 12 );
 C_ASSERT( sizeof(struct get_window_tree_request) == 16 );
 C_ASSERT( offsetof(struct get_window_tree_reply, parent) == 8 );
diff --git a/server/request_trace.h b/server/request_trace.h
index 394779c8fa4..ae521aa1764 100644
--- a/server/request_trace.h
+++ b/server/request_trace.h
@@ -1864,6 +1864,20 @@ static void dump_get_window_children_from_point_reply( const struct get_window_c
     dump_varargs_user_handles( ", children=", cur_size );
 }
 
+static void dump_get_window_from_point_request( const struct get_window_from_point_request *req )
+{
+    fprintf( stderr, " parent=%08x", req->parent );
+    fprintf( stderr, ", x=%d", req->x );
+    fprintf( stderr, ", y=%d", req->y );
+    fprintf( stderr, ", dpi=%d", req->dpi );
+}
+
+static void dump_get_window_from_point_reply( const struct get_window_from_point_reply *req )
+{
+    fprintf( stderr, " handle=%08x", req->handle );
+    fprintf( stderr, ", style=%08x", req->style );
+}
+
 static void dump_get_window_tree_request( const struct get_window_tree_request *req )
 {
     fprintf( stderr, " handle=%08x", req->handle );
@@ -3813,6 +3827,7 @@ static const dump_func req_dumpers[REQ_NB_REQUESTS] =
     (dump_func)dump_get_window_list_request,
     (dump_func)dump_get_class_windows_request,
     (dump_func)dump_get_window_children_from_point_request,
+    (dump_func)dump_get_window_from_point_request,
     (dump_func)dump_get_window_tree_request,
     (dump_func)dump_set_window_pos_request,
     (dump_func)dump_get_window_rectangles_request,
@@ -4135,6 +4150,7 @@ static const dump_func reply_dumpers[REQ_NB_REQUESTS] =
     (dump_func)dump_get_window_list_reply,
     (dump_func)dump_get_class_windows_reply,
     (dump_func)dump_get_window_children_from_point_reply,
+    (dump_func)dump_get_window_from_point_reply,
     (dump_func)dump_get_window_tree_reply,
     (dump_func)dump_set_window_pos_reply,
     (dump_func)dump_get_window_rectangles_reply,
@@ -4457,6 +4473,7 @@ static const char * const req_names[REQ_NB_REQUESTS] =
     "get_window_list",
     "get_class_windows",
     "get_window_children_from_point",
+    "get_window_from_point",
     "get_window_tree",
     "set_window_pos",
     "get_window_rectangles",
diff --git a/server/window.c b/server/window.c
index e069fc9db0e..af0eb5e8eee 100644
--- a/server/window.c
+++ b/server/window.c
@@ -1148,6 +1148,30 @@ static int all_windows_from_point( struct window *top, int x, int y, unsigned in
     return 1;
 }
 
+/* return the first window containing point (in absolute coords) */
+static struct window *first_window_from_point( struct window *top, int x, int y, unsigned int dpi )
+{
+    if (!is_desktop_window( top ) && !is_desktop_window( top->parent ))
+    {
+        screen_to_client( top->parent, &x, &y, dpi );
+        dpi = get_window_dpi( top->parent );
+    }
+
+    if (!is_point_in_window( top, &x, &y, dpi )) return NULL;
+
+    if (!(top->style & (WS_MINIMIZE|WS_DISABLED)) && point_in_rect( &top->client_rect, x, y ))
+    {
+        if (!is_desktop_window(top))
+        {
+            x -= top->client_rect.left;
+            y -= top->client_rect.top;
+        }
+        return child_window_from_point( top, x, y );
+    }
+
+    return top;
+}
+
 
 /* return the thread owning a window */
 struct thread *get_window_thread( user_handle_t handle )
@@ -2860,6 +2884,18 @@ DECL_HANDLER(get_window_children_from_point)
     else free( array.handles );
 }
 
+/* get the first window that contains a given point */
+DECL_HANDLER(get_window_from_point)
+{
+    struct window *win, *parent = get_window( req->parent );
+
+    if (!parent) return;
+    if (!(win = first_window_from_point( parent, req->x, req->y, req->dpi ))) return;
+
+    reply->handle = win->handle;
+    reply->style = win->style;
+}
+
 
 /* get window tree information from a window handle */
 DECL_HANDLER(get_window_tree)
-- 
2.55.0

